swaybg/main.c

```diff
@@ -21,6 +22,99 @@ enum scaling_mode {
 	SCALING_MODE_TILE,
 };
 
+
+#ifndef GDK_PIXBUF_CHECK_VERSION
+#define GDK_PIXBUF_CHECK_VERSION(major,minor,micro) \
+		(GDK_PIXBUF_MAJOR > (major) || \
+		 (GDK_PIXBUF_MAJOR == (major) && GDK_PIXBUF_MINOR > (minor)) || \
+		 (GDK_PIXBUF_MAJOR == (major) && GDK_PIXBUF_MINOR == (minor) && \
+			GDK_PIXBUF_MICRO >= (micro)))
+#endif
+
+cairo_surface_t *
+__gdk_pixbuf_create_cairo_surface(const GdkPixbuf *gdkbuf)
+{
+
+	int chan = gdk_pixbuf_get_n_channels(gdkbuf);
+	if (chan < 3) return NULL;
+
+	cairo_format_t fmt = (chan == 3) ? CAIRO_FORMAT_RGB24 : CAIRO_FORMAT_ARGB32;
+
+#if GDK_PIXBUF_CHECK_VERSION(2,32,0)
+	const guint8* gdkpix = gdk_pixbuf_read_pixels(gdkbuf);
+#else
+	const guint8* gdkpix = gdk_pixbuf_get_pixels(gdkbuf);
+#endif
+
+	gint w = gdk_pixbuf_get_width(gdkbuf);
+	gint h = gdk_pixbuf_get_height(gdkbuf);
+	int stride = gdk_pixbuf_get_rowstride(gdkbuf);
+	
+	
+	cairo_surface_t * cs = cairo_image_surface_create (fmt, w, h);
+	cairo_surface_flush (cs);
+	if ( !cs || cairo_surface_status (cs) != CAIRO_STATUS_SUCCESS)
+		return NULL;
+
+	int cstride = cairo_image_surface_get_stride (cs);
+	unsigned char * cpix = cairo_image_surface_get_data (cs);
+
+	if (chan == 3) {
+		int i;
+		for (i = h; i; --i) {
+			const guint8 *gp = gdkpix;
+			unsigned char *cp = cpix;
+			const guint8* end = gp + 3*w;
+			while (gp < end) {
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
+				cp[0] = gp[2];
+				cp[1] = gp[1];
+				cp[2] = gp[0];
+#else
+				cp[1] = gp[0];
+				cp[2] = gp[1];
+				cp[3] = gp[2];
+#endif
+				gp += 3;
+				cp += 4;
+			}
+			gdkpix += stride;
+			cpix += cstride;
+		}
+	} else {
+#define FIXPMUL(x,a,b) G_STMT_START { guint t = a * b + 0x80; x = (t + (t >> 8)) >> 8; } G_STMT_END
+		// premultiply alpha as fixed point 8.8
+		// [(t + 1 + ( (t + 1) >> 8)) >> 8] is approximately t div 255
+		// replaced 1 with 0x80 for rounding purposes
```

Comment:
Correct me if I'm wrong, but your rounding trick doesn't seem to work well. I tried fooling around with the formula in octave and python, and `0x80` doesn't seem to work well at all.

Some observations I've made (looking at values of `t=[1,10000000]`) is that depending on the offset you'll end up making one `2^16` interval exact:
* offset=1: `[0,2^16]` is exact, `[2^16+1,inf]` is one too small at times
* offset=2: `[0,2^16]` is one too large at times, `[2^16+1,2*2^16]` is exact, `[2*2^16+1,inf]` is one too small at times
* offset=n: `[0,(n-1)*2^16]` is one too large at times, `[(n-1)*2^16+1,n*2^16]` is exact, `[n*2^16+1,inf]` is one too small at times

In general, using a offset of `1+(t>>16)` seems to† give the exact result everywhere. So, depending on the range needed, one could either use `1` (if exact results for 16 bit unsigned is enough) or consider spending a few cycles calculating `1+(t>>16)` if one is to pedantic about correctness.

† As in it seems to be the case empirically but that I can't bother proving this formally.